### PR TITLE
GH-33987: [R] Support new dplyr .by/by argument

### LINKED ISSUE #33987: [R] Support new dplyr .by/by argument
### Describe the enhancement requested

See https://dplyr.tidyverse.org/reference/dplyr_by.html for details. 

### Component(s)

R

### PR DESCRIPTION
**Pull request by eitsupi**
### Rationale for this change

Implement the `.by` argument for `mutate`, `summarise`, `filter` and `slice_*` family.

### What changes are included in this PR?

The `.by` argument that matches `dplyr` has been added to some functions.

Most of the internal functions, such as `compute_by`, are copied from the existing `dplyr` backends, `dbplyr` and `dtplyr`.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* Closes: #33987

### CODE REVIEW

**Comment by eitsupi** [comment]
> @eitsupi - if you give this PR a rebase to fix the CI failures, I'll take a look at this

Done.

File: r/R/dplyr-by.R [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+  if (quo_is_null(by)) {
+    return(invisible(NULL))
+  }
+
+  if (is_grouped_adq(data)) {
+    message <- paste0(
+      "Can't supply `", by_arg, "` when `",
+      data_arg, "` is a grouped arrow dplyr query."
+    )
```
Comment (thisisnic) [inline r/R/dplyr-by.R:55]: The end user isn't necessarily aware of what a "grouped arrow dplyr query" is.  It's great that you're trying to mirror the original dplyr error message, but I think we might want to consider doing something else here.  Maybe "Can't supply `.by` on grouped data"?  Or even diverging from the dplyr messaging and going with something like "`.by` can only be used on ungrouped data".

**Review by thisisnic — COMMENTED** [review]
Thanks for making this PR; it's always good to keep up with the changes in the dplyr API, and this is really thorough.

Please could you add or update some of the tests to use more than 1 grouping variable here?

File: r/tests/testthat/test-dplyr-filter.R [hunk 2]
```diff
@@ -425,3 +425,34 @@ test_that("filter() with across()", {
...
+      collect(),
+    tbl
+  )
+  compare_dplyr_binding(
+    .input %>%
+      filter(int > 2, pnorm(dbl) > .99, .by = chr) %>%
+      collect(),
+    tbl,
+    warning = "Expression pnorm\\(dbl\\) > 0.99 not supported in Arrow; pulling data into R"
+  )
```
[the remark this answers is not in the document]
Comment (eitsupi) [reply r/tests/testthat/test-dplyr-filter.R:451]: This is an intentional test to ensure that no grouping occurs during conversion to data.frame.
Added a comment.

**Comment by eitsupi** [comment]
> Please could you add or update some of the tests to use more than 1 grouping variable here?

I think I did it, but these are pretty much the same test cases. I think it is better to parameterize these tests using `patrick`.
I created an issue for that #35844.

File: r/tests/testthat/test-dplyr-filter.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (thisisnic) [reply r/tests/testthat/test-dplyr-filter.R:472]: Great, thanks for clarifying that!

**Review by thisisnic — APPROVED** [review]
I'm happy with this, thanks! Will leave it a little longer before merging in case any other R folk want to have a skim over.

**Comment by eitsupi** [comment]
Are there any plans for merging? (I am just worried this has been forgotten)

**Comment by eitsupi** [comment]
Thanks for merging!

**Pull request merged**